lib/fs/casefs.go

```diff
@@ -332,85 +324,47 @@ func (r *defaultRealCaser) realCase(name string) (string, error) {
 		return out, nil
 	}
 
-	r.mut.Lock()
-	defer func() {
-		if r.count > caseMaxCachedNames {
-			select {
-			case r.timerStop <- struct{}{}:
-			default:
-			}
-			r.dropCacheLocked()
-		}
-		r.mut.Unlock()
-	}()
-
 	node := r.root
 	for _, comp := range strings.Split(name, string(PathSeparator)) {
-		if node.dirNames == nil {
-			// Haven't called DirNames yet
+		if node.dirNames == nil || node.expires.Before(time.Now()) {
+			// Haven't called DirNames yet, or the node has expired
+
 			var err error
 			node.dirNames, err = r.fs.DirNames(out)
 			if err != nil {
 				return "", err
 			}
+
 			node.dirNamesLower = make([]string, len(node.dirNames))
 			for i, n := range node.dirNames {
 				node.dirNamesLower[i] = UnicodeLowercase(n)
 			}
-			node.children = make(map[string]*caseNode)
-			node.results = make(map[string]*caseNode)
-			r.count += len(node.dirNames)
-		} else if child, ok := node.results[comp]; ok {
-			// Check if this exact name has been queried before to shortcut
-			node = child
-			out = filepath.Join(out, child.name)
-			continue
-		}
-		// Actually loop dirNames to search for a match
-		n, err := findCaseInsensitiveMatch(comp, node.dirNames, node.dirNamesLower)
-		if err != nil {
-			return "", err
+
+			node.expires = time.Now().Add(caseCacheTimeout)
+			node.child = nil
 		}
-		child, ok := node.children[n]
-		if !ok {
-			child = &caseNode{name: n}
+
+		// If we don't already have a correct cached child, try to find it.
+		if node.child == nil || node.child.name != comp {
```

Comment:
Note this also means we will not cache bad lookups (case differs from the one in the filesystem), but we don't need to optimize for those because they are a rare case and will result in an error higher up in the call chain.